### PR TITLE
Implementation of Fortran BMI Adapter and Formulation

### PR DESCRIPTION
**Pull request by robertbartel**
Implementing catchment formulations using integration with external Fortran BMI modules.

### CODE REVIEW

File: doc/BMI_MODELS.md
Comment (robertbartel) [reply doc/BMI_MODELS.md:193]: _Required Additional Fortran Registration Function_ is intended to be nested at a deeper level.  Or was this referring to something else?

File: doc/BMI_MODELS.md
Comment (robertbartel) [reply doc/BMI_MODELS.md:197]: Fixed in pending commit.

File: doc/BMI_MODELS.md
Comment (robertbartel) [reply doc/BMI_MODELS.md:208]: An adjustment to this is pending, but I'm not sure I follow exactly what you were getting at.

File: doc/BMI_MODELS.md
Comment (robertbartel) [reply doc/BMI_MODELS.md:215]: Updated in pending commit.

File: extern/iso_c_fortran_bmi/CMakeLists.txt
Comment (robertbartel) [reply extern/iso_c_fortran_bmi/CMakeLists.txt:5]: Updated in pending commit.

File: extern/iso_c_fortran_bmi/CMakeLists.txt
Comment (robertbartel) [reply extern/iso_c_fortran_bmi/CMakeLists.txt:7]: Updated in pending commit.

File: extern/iso_c_fortran_bmi/src/bmi.f90
Comment (robertbartel) [reply extern/iso_c_fortran_bmi/src/bmi.f90:299]: I don't think it is necessary.  They aren't really necessary, and their absence doesn't violate anything.  The only risk I see is potentially someone adding the "missing" forward extern declaration to the framework header file, without a backing Fortran function actually being present.  I've added (in a pending commit) a comment to _Bmi_Fortran_Common.h_ to mitigate the risk of that.

File: include/realizations/catchment/Bmi_C_Adapter.hpp
Comment (robertbartel) [reply include/realizations/catchment/Bmi_C_Adapter.hpp:127]: This leads me to a code style question: should we "require" function override declarations to include docstrings?  Or does that expectation fall on the upstream type?

This is a bit of a special case, but in principle the idea is the same:  the summary of the purpose of this function is well documented elsewhere.  Should it really be repeated here?

File: doc/BMI_MODELS.md
Comment (hellkite500) [reply doc/BMI_MODELS.md:193]: No, I didn't realize it was nested.

File: doc/BMI_MODELS.md
Comment (hellkite500) [reply doc/BMI_MODELS.md:208]: Just a minor gramar edit `integrating with C BMI module` to `integrating with a C BMI module`.  Literally just adding `a` to make it read better.

File: extern/noah-mp-modular/CMakeLists.txt
Comment (robertbartel) [reply extern/noah-mp-modular/CMakeLists.txt:7]: Updated in pending commit.

File: extern/iso_c_fortran_bmi/CMakeLists.txt
Comment (hellkite500) [reply extern/iso_c_fortran_bmi/CMakeLists.txt:5]: I meant to address these, but thanks for doing so!

File: extern/iso_c_fortran_bmi/src/bmi.f90
Comment (hellkite500) [reply extern/iso_c_fortran_bmi/src/bmi.f90:299]: We can leave it for now, shouldn't be hard to update in the future if we need to.

File: include/realizations/catchment/Bmi_C_Adapter.hpp
Comment (hellkite500) [reply include/realizations/catchment/Bmi_C_Adapter.hpp:127]: That makes total sense, just reading these diffs it wasn't clear why some had extensive doc strings and others didn't.

Comment (hellkite500) [reply include/realizations/catchment/Bmi_C_Adapter.hpp:127]: I guess the big questions is whether doxygen knows enough of the inheritance to include the parent doc strings with the virtual override in the children.

File: include/realizations/catchment/Bmi_Fortran_Adapter.hpp
Comment (robertbartel) [reply include/realizations/catchment/Bmi_Fortran_Adapter.hpp:321]: Fair point.  Updated in pending commit.

File: include/realizations/catchment/Bmi_Fortran_Adapter.hpp
Comment (robertbartel) [reply include/realizations/catchment/Bmi_Fortran_Adapter.hpp:364]: What's most important is checking that the array sizes match.  That wasn't being done, but the above mentioned change to also get the number of bytes (and then with that compare the variable array size to the vector size), should guarantee the arrays match and that individual indices contain something that should be interpreted as the right value. 

This could still be used improperly if the types don't actually agree, if they are of the same item size, but I'm not sure how much we want to worry about that.

File: include/realizations/catchment/Formulation_Constructors.hpp
Comment (robertbartel) [reply include/realizations/catchment/Formulation_Constructors.hpp:36]: Yes, that should work.  The Fortran types no longer extend the C types, so there isn't that to worry about.  And all that `BMI_C_LIB_ACTIVE` does now is add the `NGEN_BMI_C_LIB_ACTIVE` directive, which is only used here if I am not mistake (in previous revisions it was a little more significant).  But nothing in the Fortran types depends on the C types in any way.

File: include/utilities/ExternalIntegrationException.hpp
Comment (robertbartel) [reply include/utilities/ExternalIntegrationException.hpp:13]: Updated in pending commit.

File: src/core/nexus/HY_PointHydroNexusRemote.cpp
Comment (robertbartel) [reply src/core/nexus/HY_PointHydroNexusRemote.cpp:106]: Previously, this function wasn't explicitly returning anything.  That was valid, and somewhat understandable (this was the condition when the output should be sent to a remote instead of written locally), but it resulted in a warning that was contributing to lots of compiler output noise.  It also happened to be very close in proximity to how the output is handled, so it made sense to use a specific value both for improving general internal operation and to help debug the issue of duplicate writes that we are seeing.

File: include/realizations/catchment/Bmi_Fortran_Adapter.hpp
Comment (hellkite500) [reply include/realizations/catchment/Bmi_Fortran_Adapter.hpp:364]: I agree, the important part is making sure the array sizes match.  After noticing that, I got hung up thinking about how that extends to the at indicices functions.

File: src/realizations/catchment/Bmi_Fortran_Adapter.cpp
Comment (robertbartel) [reply src/realizations/catchment/Bmi_Fortran_Adapter.cpp:200]: Updated in pending commit.

File: test/realizations/catchments/Bmi_C_Adapter_Test.cpp
Comment (robertbartel) [reply test/realizations/catchments/Bmi_C_Adapter_Test.cpp:136]: I think that was used in several places to provide more control over the formatting, though at this point I'm not exactly sure.  But those weren't modified as part of this change set.

File: doc/BMI_MODELS.md
Comment (robertbartel) [reply doc/BMI_MODELS.md]: Ah, I see.  I've address that also.

File: extern/iso_c_fortran_bmi/CMakeLists.txt
Comment (hellkite500) [inline extern/iso_c_fortran_bmi/CMakeLists.txt:6]: The debug flags should be appropriately added only to debug build types, with something like the following generator expression.  This addes `-Wall` to all builds, but only the extra debug flags to the Debug build.
```cmake
add_compile_options(
  "-Wall"
  "$<$<CONFIG:DEBUG>:-g;-fbacktrace;-fbounds-check>"
)
```

**Review by hellkite500 — APPROVED** [review]

**Comment by hellkite500** [comment]
I think what we have here is in pretty good shape.  The only conversation left to have is what we want to do about testing.

**Comment by robertbartel** [comment]
@hellkite500, FYI, I rebased after another PR got merged to `master`, so this will have to be re-approved  (though that can wait if it needs to, while the overall testing strategy is discussed).

**Comment by hellkite500** [comment]
Will add unit testing support of BMI Fortran adapter before merging.  This requires writing a fortran BMI model, and writing the unit tests that excercise that model via the adapter (and hence indirectly the BMI ISO C interface).  This will be tracked by #290 and incrementally added to this PR.

**Comment by hellkite500** [comment]
Testing in #290 complete.  GH actions have been updated to run the fortran tests independently.   We decided to ensure the Fortran Formulation tests were enabled and functioning as the final piece of this PR.  Those tests are nearly complete, there were 4 previously `DISABLED` tests that we want to get working for the Fortran formulation.

File: extern/iso_c_fortran_bmi/CMakeLists.txt
Comment (hellkite500) [reply extern/iso_c_fortran_bmi/CMakeLists.txt:6]: fixed with 2391f54

File: extern/iso_c_fortran_bmi/README.md
Comment (donaldwj) [inline extern/iso_c_fortran_bmi/README.md:15]: Is this still accurate> I remember you saying that you removed the dependency on 'save'

File: extern/iso_c_fortran_bmi/src/bmi.f90
Comment (donaldwj) [inline extern/iso_c_fortran_bmi/src/bmi.f90:71]: Are we certian that we don't need the 'set_value_at_indices()' functions

File: extern/iso_c_fortran_bmi/test/test_iso_c.c
Comment (donaldwj) [inline extern/iso_c_fortran_bmi/test/test_iso_c.c:47]: possibly move these to a header file.

File: extern/noah-mp-modular/CMakeLists.txt
Comment (donaldwj) [inline extern/noah-mp-modular/CMakeLists.txt:36]: we should at the very least try to read this variable from the shell before using a hard coded value (which is wrong for many distributions)

File: extern/iso_c_fortran_bmi/README.md
Comment (hellkite500) [reply extern/iso_c_fortran_bmi/README.md:15]: I just pushed an update to the register function, I'll fix this documentation real quick.

File: extern/test_bmi_fortran/src/bmi.f90
Comment (donaldwj) [inline extern/test_bmi_fortran/src/bmi.f90:1]: This looks like a file used durring debugging as almost all of the bmi interface is disabled. Should this be in the pull request and if so is it up to date?

File: src/NGen.cpp
Comment (donaldwj) [inline src/NGen.cpp:133]: Do we want to keep this in master?

File: src/realizations/catchment/Bmi_Fortran_Formulation.cpp
Comment (donaldwj) [inline src/realizations/catchment/Bmi_Fortran_Formulation.cpp:146]: I have seen models where negative time steps where real I think it was in DFLOW. Do we want to account for this possibility?

**Review by donaldwj — DISMISSED** [review]
There were a few minor issues which I have created conversations for. None of the issues would require waiting on merging this.

File: extern/iso_c_fortran_bmi/src/bmi.f90
Comment (hellkite500) [reply extern/iso_c_fortran_bmi/src/bmi.f90:71]: We don't need them for the immediate need of using NoahMP fortran module.  We can extend this library with implementations for the unimplemented functions in the future.  Ideally, we should proxy each BMI function, and then any model that doesn't want to implement a particular function will still need to define it and simply always return `BMI_FAILURE`.

File: extern/iso_c_fortran_bmi/test/test_iso_c.c
Comment (hellkite500) [reply extern/iso_c_fortran_bmi/test/test_iso_c.c:47]: In the core NGEN framework code, they are in a header, this was just meant to be a standalone C/Fortran test to exercise the functions.  Not sure it needs much attention beyond that.

File: extern/noah-mp-modular/CMakeLists.txt
Comment (hellkite500) [reply extern/noah-mp-modular/CMakeLists.txt:36]: This snuck in, I'll remove it.  The find netcdf cmake function doesn't seem to work correctly on some systems, so this was hard coded in for that specific instance.

File: extern/test_bmi_fortran/src/bmi.f90
Comment (hellkite500) [reply extern/test_bmi_fortran/src/bmi.f90:1]: It doesn't have to be for the purpose of the test model and the use of the ISO C library.  But to build the test model "stand alone" it would be required.  I forgot to uncomment the interface when I was debugging the module issue.  Ideally, the BMI dependency would be managed entirely out of source, but in cases where a model "customizes" the implementation of BMI by ignoring some functions (via commenting out the code) then it is required to carry around the definition.  I pushed a change to uncommment the relevant interface functions to make it useful, but I could easily be convinced to drop it from the PR.  Any thoughts?

File: src/NGen.cpp
Comment (hellkite500) [reply src/NGen.cpp:133]: @robertbartel what are your thoughts?  When I first seen it, I thought about putting hiding some of this in a function in a utility header and then using a `DEBUG LEVEL` macro to call that function if needed.

File: src/realizations/catchment/Bmi_Fortran_Formulation.cpp
Comment (hellkite500) [reply src/realizations/catchment/Bmi_Fortran_Formulation.cpp:146]: Not in the scope of this PR 😄 .  It might be worth considering in an issue to relax the requirement in the future.

**Comment by hellkite500** [comment]
> There were a few minor issues which I have created conversations for. None of the issues would require waiting on merging this.

I do want to push a documentation update and one small BMI_Fortran_Adapter modification before this is merged.  Will have those done shortly.

File: extern/iso_c_fortran_bmi/README.md
Comment (hellkite500) [reply extern/iso_c_fortran_bmi/README.md]: updated with a0ca7e9fe21cd5c120d2be04331d6e6c23b62796

File: src/NGen.cpp
Comment (robertbartel) [reply src/NGen.cpp:133]: My thought is leave it in.  First, it's a very small amount of code being added (or not).  Second, the purpose was to add support for "remote" debugging by way of a Docker container, to hopefully better be able to step through the Fortran code.  The setup just wasn't finished before the issues needing debugging were otherwise resolved. Still, I think it's good to leave this in place, rather than lose the progress (the general idea came from something I started trying to debug the MPI related work).

As far as the utility function, I think this is minimal enough to just leave where it is, unless you anticipate something else needing to reuse the functionality.   I also think it should have a dedicated macro, rather than be tied directly to a debug level, for more granular control over whether this is active.

File: extern/noah-mp-modular/CMakeLists.txt
Comment (hellkite500) [reply extern/noah-mp-modular/CMakeLists.txt:36]: This is turning out to be tricky getting NetCDF to link correctly in a generic cmake file.

File: src/NGen.cpp
Comment (hellkite500) [reply src/NGen.cpp:133]: I'm OK with it, just like to make sure anything going in the "main" source/function is thought through a bit.

**Review by hellkite500 — APPROVED** [review]
I approve this PR to merge, though will wait for an extra external reviewer.  An issue with the NoahMP cmake build still (potentially) exists trying to get the shared library to build correctly and link with NetCDF.  Propose handling that in its own issue/PR around supporting NoahMP, not the generic Fortran support.

File: extern/noah-mp-modular/CMakeLists.txt
Comment (donaldwj) [inline extern/noah-mp-modular/CMakeLists.txt:10]: So that is how you add flags to jus the debug build

**Review by donaldwj — APPROVED** [review]
I dont see any issues with the final changes

**Pull request merged**
